packages/cicero-core/src/instanceloader.js

```diff
@@ -0,0 +1,205 @@
+        properties.map((property) => property.getDecorators().map((decorator) => {
+            if (decorator.getName() === 'ContractParty') {
+                const data = instance.data;
+                const partyName = data[property.name];
+                instance.parties.push(partyName);
+            }
+        }));
+        return instance;
+    }
+
+    /**
+     * Create an instance from a Template.
+     * @param {*} Instance - the type to construct
+     * @param {Template} template  - the template for the instance
+     * @return {object} - the contract instance
+     */
+    static fromTemplate(Instance, template) {
+        const templateMetadata = template.getMetadata();
+        const logicManager = template.getLogicManager();
+        const parserManager = new ParserManager(logicManager.getModelManager(), null, this.instanceKind);
+        const ciceroMarkTransformer = new CiceroMarkTransformer();
+        const ergoEngine = new ErgoEngine();
+        const input = templateMetadata.getSample();
+        const grammar = template.getParserManager().getTemplate();
+
+        // Initialize the parser
+        Util.initParser(
+            parserManager,
+            logicManager,
+            ergoEngine,
+            templateMetadata.getIdentifier(),
+            templateMetadata.getTemplateType(),
+            grammar,
+            templateMetadata.getRuntime(),
+        );
+
+        const data = Util.parseText(parserManager, ciceroMarkTransformer, input, null, null, 'text/sample.md');
+        return InstanceLoader.fromTemplateWithData(Instance, template, data);
+    }
+
+    /**
+     * Create an instance from an archive
+     * @param {*} Instance - the type to construct
+     * @param {Buffer} buffer  - the buffer to a Smart Legal Contract (slc) file
+     * @param {object} options - additional options
+     * @return {Promise<Instance>} a Promise to the instance
+     */
+    static async fromArchive(Instance, buffer, options) {
+        const logicManager = new LogicManager('es6', null, options);
+
+        const method = 'fromArchive';
+        const zip = await JSZip.loadAsync(buffer);
+        const ctoModelFiles = [];
+        const ctoModelFileNames = [];
+
+        // add metadata
+        const packageJsonObject = await InstanceLoader.loadZipFileContents(zip, 'package.json', true, true);
+        // XXX likely we will need to discuss what metadata is relevant / needed for instances
+        // XXX For now hijack the the template metadata
+        const metadata = new InstanceMetadata(
+            packageJsonObject,
+        );
+
+        // add contract data
+        const data = await InstanceLoader.loadZipFileContents(zip, 'data.json', true, true);
+
+        // add contract history
+        const history = await InstanceLoader.loadZipFileContents(zip, 'history.json', true, true);
+
+        //grab instantiator
+        const instantiator = history.length !== 0 ? history[0].currentState.instatiator : null;
```

Comment:
Can `history[0].currentState` be null?

Reply:
No. The `previousState` would be `null` when the contract is first instantiated but the `currentState` would not.